Add build config files for Ray example pipelines

tscurtu opened the pull request:
This is a follow-up to #844 that adds build configuration files to the example pipelines which require `Ray` as a dependency.

This is enabled by adding the source code of the example pipelines which depend on ray to the "extra" wheels.

## Testing

Because these pipelines now depend on the version of `sematic[ray]` that is built using these changes, which are not released to PyPi, they were tested using a local build of the Sematic wheel. [This commit](https://github.com/sematic-ai/sematic/pull/857/commits/46ae583f0d0656f77e0dfbe8f027ba1e43776418) contains the version that was tested.

**Example Testing Pipeline** run command:

```
$ sematic run --build sematic/examples/testing_pipeline/__main__.py -- --images --cloud --detach
```

Resulting Dashboard entry: [link](https://dev1.dev-usw2-sematic0.sematic.cloud/runs/8813fdd626c042c8bf03e1500cf884e8#tab=output&panel=run&run=3241e3f3628d455b8fe8459380d1c423).

**Example Lightning ResNet** run command:

```
$ sematic run --build sematic/examples/lightning_resnet/main.py -- --cloud
```

Resulting Dashboard entry: [link](https://dev1.dev-usw2-sematic0.sematic.cloud/runs/1f75060f92bf40dd91f12eb042af221e#panel=run&tab=logs&run=0ed5f1179b144a629e8e7faf09c78051) (did not complete due to [this error](https://github.com/sematic-ai/sematic/pull/857#issuecomment-1572144611))

**Example CIFAR10 Classifier** run command:

```
$ sematic run --build sematic/examples/cifar_classifier/main.py -- --cloud
```

Resulting Dashboard entry: [link](https://dev1.dev-usw2-sematic0.sematic.cloud/runs/3f02586fd29a4839adfbbe36d9f5edf5#panel=run&tab=logs&run=34c16ddf0e0746188554caa6a047ba68) (did not complete due to [this error](https://github.com/sematic-ai/sematic/pull/857#issuecomment-1572144611))

augray commented on sematic/ee/BUILD:
```diff
@@ -2,8 +2,12 @@ sematic_py_lib(
+        "//sematic:init",
         "//sematic/ee/plugins/external_resource/ray:checkpoint",
         "//sematic/ee/plugins/external_resource/ray:cluster",
+        "//sematic/examples/cifar_classifier:cifar_classifier_lib",
```
It seems strange to say that `ray` depends on these examples, rather than the other way around. If you just want to make sure these examples get included in the wheel, modify `extras.bzl` so that `EXTRA_SEMATIC_DEPS` gets these as dependencies.

augray commented on sematic/ee/BUILD:
```diff
@@ -2,8 +2,12 @@ sematic_py_lib(
+        "//sematic:init",
         "//sematic/ee/plugins/external_resource/ray:checkpoint",
         "//sematic/ee/plugins/external_resource/ray:cluster",
+        "//sematic/examples/cifar_classifier:cifar_classifier_lib",
```
It seems strange to say that `ray` depends on these examples, rather than the other way around. If you just want to make sure these examples get included in the wheel, modify `extras.bzl` so that `EXTRA_SEMATIC_DEPS` gets these as dependencies.

tscurtu replied:
Done.

augray commented on sematic/examples/cifar_classifier/BUILD:
```diff
@@ -6,10 +6,7 @@ sematic_py_lib(
-        "//sematic:init",
-        "//sematic/ee:ray",
-    ],
+    deps = [],
```
Again, this seems backwards.

augray commented on sematic/examples/lightning_resnet/BUILD:
```diff
@@ -7,10 +7,7 @@ sematic_py_lib(
-        "//sematic:init",
-        "//sematic/ee:ray",
-    ],
+    deps = [],
```
same comment as above

tscurtu replied:
Done.

tscurtu commented:
Hitting an issue where the Ray Agent dies on the Ray Workers with no real debugging information. The Ray pods continue in zombie mode forever.

Ray Worker info:
```
root@rc0547849f186b8d5ea42ecb94f-worker-worker-group-0-6rnwx:/# cat /tmp/ray/session_2023-06-01_13-42-32_415121_1/logs/raylet.err
[2023-06-01 13:43:05,421 E 26 94] (raylet) agent_manager.cc:135: The raylet exited immediately because the Ray agent failed. The raylet fate shares with the agent. This can happen because the Ray agent was unexpectedly killed or failed. Agent can fail when
- The version of `grpcio` doesn't follow Ray's requirement. Agent can segfault with the incorrect `grpcio` version. Check the grpcio version `pip freeze | grep grpcio`.
- The agent failed to start because of unexpected error or port conflict. Read the log `cat /tmp/ray/session_latest/logs/dashboard_agent.log`. You can find the log file structure here https://docs.ray.io/en/master/ray-observability/ray-logging.html#logging-directory-structure.
- The agent is killed by the OS (e.g., out of memory).
root@rc0547849f186b8d5ea42ecb94f-worker-worker-group-0-6rnwx:/#
root@rc0547849f186b8d5ea42ecb94f-worker-worker-group-0-6rnwx:/# cat /tmp/ray/session_latest/logs/dashboard_agent.log
cat: /tmp/ray/session_latest/logs/dashboard_agent.log: No such file or directory
```

Ray Head info:
```
2023-06-01 13:43:24,205 INFO (monitor) autoscaler.py:466 -- The autoscaler took 0.001 seconds to complete the update iteration.
[2023-06-01 13:43:24,647 W 116 116] (raylet) agent_manager.cc:115: Agent process expected id 424238335 timed out before registering. ip , id 0
[2023-06-01 13:43:24,650 I 116 195] (raylet) agent_manager.cc:131: Agent process with id 424238335 exited, exit code 0. ip . id 0
[2023-06-01 13:43:24,650 E 116 195] (raylet) agent_manager.cc:135: The raylet exited immediately because the Ray agent failed. The raylet fate shares with the agent. This can happen because the Ray agent was unexpectedly killed or failed. Agent can fail when
- The version of `grpcio` doesn't follow Ray's requirement. Agent can segfault with the incorrect `grpcio` version. Check the grpcio version `pip freeze | grep grpcio`.
- The agent failed to start because of unexpected error or port conflict. Read the log `cat /tmp/ray/session_latest/logs/dashboard_agent.log`. You can find the log file structure here https://docs.ray.io/en/master/ray-observability/ray-logging.html#logging-directory-structure.
- The agent is killed by the OS (e.g., out of memory).
[2023-06-01 13:43:24,650 I 116 116] (raylet) main.cc:301: Raylet received SIGTERM, shutting down...
[2023-06-01 13:43:24,650 I 116 116] (raylet) accessor.cc:435: Unregistering node info, node id = 3dc2eef92b42a949fc9ae3319dc63f3aada2094b400313dbe6178f4a
[2023-06-01 13:43:24,650 E 116 195] (raylet) agent_manager.cc:135: The raylet exited immediately because the Ray agent failed. The raylet fate shares with the agent. This can happen because the Ray agent was unexpectedly killed or failed. Agent can fail when
- The version of `grpcio` doesn't follow Ray's requirement. Agent can segfault with the incorrect `grpcio` version. Check the grpcio version `pip freeze | grep grpcio`.
- The agent failed to start because of unexpected error or port conflict. Read the log `cat /tmp/ray/session_latest/logs/dashboard_agent.log`. You can find the log file structure here https://docs.ray.io/en/master/ray-observability/ray-logging.html#logging-directory-structure.
- The agent is killed by the OS (e.g., out of memory).
[2023-06-01 13:43:24,650 I 13 13] (gcs_server) gcs_node_manager.cc:88: Draining node info, node id = 3dc2eef92b42a949fc9ae3319dc63f3aada2094b400313dbe6178f4a
[2023-06-01 13:43:24,650 I 13 13] (gcs_server) gcs_node_manager.cc:212: Removing node, node id = 3dc2eef92b42a949fc9ae3319dc63f3aada2094b400313dbe6178f4a, node name = 10.42.9.228
[2023-06-01 13:43:24,650 I 13 13] (gcs_server) gcs_placement_group_manager.cc:739: Node 3dc2eef92b42a949fc9ae3319dc63f3aada2094b400313dbe6178f4a failed, rescheduling the placement groups on the dead node.
[2023-06-01 13:43:24,650 I 13 13] (gcs_server) gcs_actor_manager.cc:991: Node 3dc2eef92b42a949fc9ae3319dc63f3aada2094b400313dbe6178f4a failed, reconstructing actors.
[2023-06-01 13:43:24,650 I 116 116] (raylet) io_service_pool.cc:47: IOServicePool is stopped.
[2023-06-01 13:43:24,661 I 13 13] (gcs_server) gcs_node_manager.cc:134: Raylet 3dc2eef92b42a949fc9ae3319dc63f3aada2094b400313dbe6178f4a is drained. Status GrpcUnavailable: RPC Error message: Cancelling all calls; RPC Error details: . The information will be published to the cluster.
2023-06-01 13:43:29,208 INFO (monitor) autoscaler.py:147 -- The autoscaler took 0.0 seconds to fetch the list of non-terminated nodes.
2023-06-01 13:43:29,208 INFO (monitor) load_metrics.py:161 -- LoadMetrics: Removed ip: 3dc2eef92b42a949fc9ae3319dc63f3aada2094b400313dbe6178f4a.
2023-06-01 13:43:29,208 INFO (monitor) load_metrics.py:164 -- LoadMetrics: Removed 1 stale ip mappings: {'3dc2eef92b42a949fc9ae3319dc63f3aada2094b400313dbe6178f4a'} not in set()
2023-06-01 13:43:29,208 INFO (monitor) autoscaler.py:423 --
```

The only difference I can find between the Bazel and Docker versions of the pods is the size of the image, with the Docker version being ~5.6GB compared to Bazel's 1.3GB. Will pause this effort and continue again after optimizing the image size.

neutralino1 reviewed (approved):
Thank you
